feat #9: 맛집 목록 조회 API - 거리순, 평점순 정렬 구현

mojh7 opened the pull request:
## 📃 설명

- resolves #9
- 

## 🔨 작업 내용

1. 맛집 목록 조회 API 거리순 정렬 구현
2. 맛집 목록 조회 API 평점순 정렬 구현 
3. 테스트 작성
   - 0.1km 이내 없을 때 비어있는 list 반환
   - 과제 요구사항에 예시로 작성된 1.0km, 5.0km, 10.0km 이내 조회 테스트

## 💬 기타 사항

- query parameter 에 대한 유효성 검증이 이번 PR에는 없고 다음 PR에서 추가하고 잘못 된 입력으로 조회 실패하는 테스트 작성하겠습니다.
- TODO: matzip_row table 필드 맞추기
  - 위도, 경도, 음식점 이름을 matzip_row table에서 가져와야 돼서 임시로 matzip_row 엔티티 클래스 작성
  - pk에 대한 것 해당 repo에서 따로 브랜치 들어가서 보려고했는데 아직 로컬에서 작업중이셔서 회의 후 matzip_row 클래스 맞추기

- `@Sql` 이 테스트 마다 실행돼서 여러 개의 테스트 중 첫 번째 말고는 전부 실패
  - 로컬에서 matzip 데이터만 있을 때는 실패 안 했는데 matzip_row 데이터 추가하니 실패
  - 아마 fk 관련되서 실패하는 것 같고 임시로 boforeAll 에서 수동으로 로딩하도록 처리
  - https://choichumji.tistory.com/173

- 맛집 조회 타입 enum 관련 참고 사이트
  - [함수형 인터페이스](https://velog.io/@agpine12/%EB%8D%94-%EC%9E%90%EB%B0%94-Java8-01.-%ED%95%A8%EC%88%98%ED%98%95-%EC%9D%B8%ED%84%B0%ED%8E%98%EC%9D%B4%EC%8A%A4%EC%99%80-%EB%9E%8C%EB%8B%A4)

  - [자바 기본 함수형 인터페이스](https://bcp0109.tistory.com/313)

  - [자바 enum과 람다](https://kukim.tistory.com/68)

piopoi committed:
refactor #2: 공통 exception 처리 세팅

piopoi committed:
refactor #2: 사용자 로그인 요청(JWT) API 추가

- 인증/인가 관련 패키지 생성: `auth`
- validate는 컨트롤러 레이어의 책임으로 함.
  - MemberServiceTest의 invalid case 제거

piopoi committed:
refactor #2: jwt.token에서 token이 중복되는 네이밍 변경

piopoi committed:
refactor #2: bearer 매직넘버 제거

piopoi committed:
feat #4: 사용자 설정 업데이트 API 구현

mojh7 committed:
feat #9: 예외 처리 관련 클래스 추가

mojh7 committed:
feat #9: 예외 처리 관련 클래스 추가

mojh7 committed:
setup #9: log폴더 ignore 추가

mojh7 committed:
setup #9: log폴더 ignore 추가

mojh7 committed:
setup #9: log폴더 ignore 추가

mojh7 committed:
feat #9: 맛집 목록 거리 순, 평점 순 정렬 조회 구현

mojh7 committed:
feat #9: 맛집 목록 거리 순, 평점 순 정렬 조회 구현

mojh7 committed:
feat #9: 맛집 목록 거리 순, 평점 순 정렬 조회 구현

mojh7 committed:
test #9: 맛집 목록 거리 순, 평점 순 정렬 조회 테스트 작성

mojh7 committed:
test #9: 맛집 목록 거리 순, 평점 순 정렬 조회 테스트 작성

mojh7 committed:
test #9: 맛집 목록 거리 순, 평점 순 정렬 조회 테스트 작성

piopoi committed:
setup #2: jwt secret-key 제거

piopoi committed:
Merge pull request #12 from Wanted-Pre-Onboarding-Backend7-R/feat/#2-api-member-login-jwt-auth

feat #2: 사용자 로그인(JWT) API 구현

piopoi committed:
refactor #4: 위도, 경도 자료형 변경(BigDecimal -> String)

piopoi committed:
fix #4: 사용자 생성자 파라미터 수정

piopoi committed:
Merge remote-tracking branch 'origin/develop' into feat/#4-api-user-update-location

JoonheeJeong commented on src/test/java/com/wanted/teamr/tastyfinder/api/matzip/service/MatzipServiceMatzipListRetrieveTest.java:
```diff
@@ -0,0 +1,263 @@
+        // when
+        List<MatzipSummaryReponse> result = matzipService.retrieveMatzipList(request);
+
+        // then
+        assertAll(
+                () -> assertThat(result.size()).isEqualTo(2),
+                () -> assertThat(result).extracting(MatzipSummaryReponse::getName)
+                                        .containsExactlyElementsOf(expected)
+        );
```
굳이 assertAll 쓰시는 이유가 있을까요? 꺼내도 될 것 같은데 궁금해서 여쭤봅니다!

mojh7 replied:
https://velog.io/@ki960213/assertThat-vs-assertAll-vs-assertSoftly

둘의 차이가 여러 개의 assertThat 중 2개 이상이 실패할 때 assertThat만 쓰면 실패한 assertThat에서 바로 멈추고 다른 assertThat은 실행하지 않고 실패한 assertThat 1개에만 에러 메시지를 출력하고
assertAll은 여러 개 중 1개가 실패해도 그 뒤에 assertThat까지 다 테스트해보고 실패한 것만 보여줍니다

assertThat의 갯수가 한 7개, 13개, 그 이상 많을 때 한 번 테스트 실행하고 틀린 곳 하나 고치고
다시 실행해서 틀린 곳 하나 발견하고 하나 고치고 하는 것보다
assertAll은 지금 작성한 것 중에서 어떤 것들이 틀렸는지 한 번에 확인 할 수 있는 장점이 있습니다.

하지만 지금은 그냥 복수 개의 assertThat있으면 아무 생각 없이 묶었었네요
개수가 2~3개 정도 적을 때는 뭐가 좋은지 모르겠네요

assertThat
``` java
@Test
void temp() {
    assertThat(1).isEqualTo(1);
    assertThat(2).isEqualTo(2);
    assertThat(3).isEqualTo(1); // 실패
    assertThat(4).isEqualTo(1);
    assertThat(1).isEqualTo(1);
    assertThat(6).isEqualTo(1);
}
```
출력
expected: 1
 but was: 3
org.opentest4j.AssertionFailedError: 
expected: 1
 but was: 3
 
 ---
 assertAll 쓸 때
 
 ``` java
@Test
void temp() {
    assertAll(
        () -> assertThat(1).isEqualTo(1),
        () -> assertThat(2).isEqualTo(2),
        () -> assertThat(3).isEqualTo(1),
        () -> assertThat(4).isEqualTo(1),
        () -> assertThat(1).isEqualTo(1),
        () -> assertThat(6).isEqualTo(1)
    );
}
```
 출력

 Multiple Failures (3 failures)
	org.opentest4j.AssertionFailedError: 
expected: 1
 but was: 3
	org.opentest4j.AssertionFailedError: 
expected: 1
 but was: 4
	org.opentest4j.AssertionFailedError: 
expected: 1
 but was: 6
org.gradle.internal.exceptions.DefaultMultiCauseException: Multiple Failures (3 failures)
	org.opentest4j.AssertionFailedError: 
expected: 1
 but was: 3
	org.opentest4j.AssertionFailedError: 
expected: 1
 but was: 4
	org.opentest4j.AssertionFailedError: 
expected: 1
 but was: 6

JoonheeJeong replied:
오호~ 검증할 내용이 많을 때 중간에 틀려도 다 검증을 진행해서 어떤 것이 틀렸는지 골라서 확인할 수 있군요! 덕분에 지식이 늘었네요 ㅎㅎ 감사합니다!

SteadyFlow736 replied:
오! 이것은 정말 좋은 정보네요. 두 분께 감사드립니다. 🙏

JoonheeJeong commented on src/test/java/com/wanted/teamr/tastyfinder/api/matzip/controller/MatzipControllerMatzipListRetrieveTest.java:
```diff
@@ -0,0 +1,295 @@
+                                     .andReturn()
+                                     .getResponse()
+                                     .getContentAsString(StandardCharsets.UTF_8);
+
+        List<MatzipSummaryReponse> resMatzipList = objectMapper.readValue(responseBody, new TypeReference<>() {});
+        assertThat(resMatzipList).extracting(MatzipSummaryReponse::getName)
+                                 .containsExactlyElementsOf(expected);
+    }
```
1. 타입 추론도 있군요! 첨봤어요!!
2. 리스트 비교 아주 깔끔하네요!! 👍🏼

mojh7 replied:
저도 처음 봤습니다 그리고 원래는

`List<MatzipSummaryReponse> resMatzipList = objectMapper.readValue(responseBody, new TypeReference<List<MatzipSummaryReponse>>() {});`

뒤에도 타입 써줘야 되는 것 같은데
아래 리스트 작성할 때 처럼 인텔리제이에서 TypeReference내에 타입 적은 부분이 회색 표시로 되고 지워도 작동 하더군요
`List<String> example = new ArrayList<>();`

JoonheeJeong reviewed (commented):
장현님, 코드 잘 봤습니다!

SteadyFlow736 committed:
feature #13: 시군구 데이터 Redis 적재 기능 구현

SteadyFlow736 commented on src/main/java/com/wanted/teamr/tastyfinder/api/matzip/domain/MatzipListRetrieveType.java:
```diff
@@ -0,0 +1,41 @@
+                                                          .sorted(Comparator.comparingDouble(MatzipSummaryReponse::getAvgRating).reversed())
+                                                          .collect(Collectors.toList()));
+
+    private TriFunction<List<Matzip>, Location, Double, List<MatzipSummaryReponse>> retrieveFunc;
```
enum 타입의 필드 값으로 이렇게 FunctionalInterface를 이용하는 것은 생각도 못했네요. 다형성을 이렇게도 적용할 수 있는지 새롭게 알게 되었습니다. 👍

mojh7 replied:
enum에도 함수를 필드로 가질 수 있는거랑 enum으로 다형성 구현하는 것 예전에 어렴풋이 봤던 것 같은데
이번에 다시 찾아서 해봤는데 잘 작동하니 기분이 좋네요 😄

원래는 보통 인터페이스, 상속, 구현체 이런 것들로도 다형성 구현이 될텐데
목록 조회 api에서 request param으로 받은 변수값으로 로직을 구분해야되니
결국 if-else나 switch case를 쓰긴 해야되더라구요
그래서 MatzipService에서 바로 if-else 쓰기 보단 MatzipRetrieveService 라는 클래스만들거나 아니면 `@Service` 가 아니더라도 따로 클래스 하나 더 만들어서 그 클래스에서 if-else로 enum 타입 구분해서 로직을 한 번 감추려고 했습니다. 그러다가 어찌저찌 해서 enum으로 했습니다

SteadyFlow736 commented on src/main/java/com/wanted/teamr/tastyfinder/api/matzip/domain/MatzipListRetrieveType.java:
```diff
@@ -0,0 +1,41 @@
+        this.retrieveFunc = retrieveFunc;
+    }
+
+    public List<MatzipSummaryReponse> retrieve(List<Matzip> matzipList, Location requestLocation, double range) {
+        return retrieveFunc.apply(matzipList, requestLocation, range);
+    }
```
바깥에서는 FunctionalInterface::apply 대신 의미가 있는 retrieve가 호출되도록 한 단계 감쌌군요. 좀 더 의미있는 이름이 된 것 같습니다.

mojh7 replied:
다른 사람들 코드 예제 보다가 찾았는데 이름 의미있는 것으로 지정하니 좋아보이더라구요
아마 람다로 넣어서 apply 로 되어있는 것 같은데 필드를 람다 말고 일반 함수로 가지게 하면 바로 이름 호출 될 것 같습니다

그리고 지금 중식, 일식, 카페같이 음식점 종류로 필터링해서 조회하는 것 추가할 예정인데
이렇게 하려면 결국 음식점 종류 정보를 파라미터로 넣어야되고 대응되는 함수형 인터페이스도 수정해야됩니다 그래서 request dto 유효성 검사만 추가하고 람다 말고 일반적인 함수로 바꿀 것 같습니다...

SteadyFlow736 commented on src/main/java/com/wanted/teamr/tastyfinder/api/matzip/controller/MatzipController.java:
```diff
@@ -0,0 +1,26 @@
+    private final MatzipService matzipService;
+
+    @GetMapping("/api/matzips")
+    public ResponseEntity<List<MatzipSummaryReponse>> retrieveMatzipList(@ModelAttribute @Valid MatzipListRetrieveRequest request) {
```
이 부분이 좀 신기합니다. 분명 type이 String으로 들어올텐데, 이것을 MatzipListRetrieveType enum 타입으로 매핑해 주는군요. 스프링이 내부적으로 type conversion이 강력한 것 같습니다. 이런것도 사실 이렇게 누군가가 사용하는 모습을 보지 못하면 알아차리기 힘든 기능 중 하나가 아닌가 싶습니다. 이번에는 운이 좋게도 장현님 코드로 배우게 되었네요. 😀

mojh7 replied:
request param 여러 개 받고 조건에 맞게 조회하는 api 안 해봐서 혼자했으면 시간이 좀 더 걸렸을 텐데

저도 저번 주 했던 과제에 목록 조회 코드들 보면서 많이 배우고 비교적 쉽게 적용했네요

그리고 주로 request body만 썼을 때도 기껏해야 String, Long 이런 걸로만 데이터 받았는데
enum으로 해도 자동으로 매핑해주는 것 이번에 처음 알았네요 :)

SteadyFlow736 commented on src/main/java/com/wanted/teamr/tastyfinder/api/matzip/service/MatzipService.java:
```diff
@@ -0,0 +1,24 @@
+        Location requestLocation = Location.of(request.getLat(), request.getLon());
+        double range = Double.parseDouble(request.getRange());
+        return request.getType()
+                      .retrieve(matzipRepository.findAll(), requestLocation, range);
```
오 다형성! 👍

SteadyFlow736 reviewed (approved):
코드를 다양한 방면에서 깊게 고민하시고 좋은 방법으로 작성하고 계시는 것 같습니다.  수고 많으셨습니다. 👍

SteadyFlow736 committed:
docs #13: 코멘트 추가

SteadyFlow736 committed:
refactor #13: SggDataLoadService 리팩토링

SteadyFlow736 committed:
test #13: SggDataLoadService 통합테스트 추가

SteadyFlow736 committed:
test #13: SggDataService 통합테스트 추가

SteadyFlow736 committed:
test #13: SggData 테스트 추가

SteadyFlow736 committed:
docs #13: SggData 코멘트 추가

piopoi committed:
feat #4: Global Exception 처리 추가

mojh7 committed:
setup #20: PR 바로 생성, draft -> ready for review 전환 시 디스코드 알림 설정

 - discord webhook url은 github secret에 저장

mojh7 committed:
setup #20: yaml 파일 잘 못된 것 올렸어서 다시 수정

mojh7 committed:
Merge pull request #22 from Wanted-Pre-Onboarding-Backend7-R/setup/#20-pr-discord-notification

setup #20: PR 추가 시 Discord 알림 설정

piopoi committed:
feat #4: 인증/인가 exception handler 추가

- 인증 exception handler: AuthenticationEntryPointImpl
- 인가 exception handler: AccessDeniedHandlerImpl

mojh7 committed:
fix #9: 충돌 해결

mojh7 committed:
fix #9: 충돌 해결

mojh7 committed:
fix #9: 충돌 해결

mojh7 committed:
setup #24: notify-pr.yml 수정

 - color랑 content 수정

mojh7 committed:
setup #24: pr closed 때 merge로 된거면 알림

mojh7 committed:
setup #24: auto labeling

 - issue도 되지만 issue template으로 생성할 때 자동으로 붙으니 false
 - pr 제목만 체크하길 바랬지만 내용도 체크하는듯...

mojh7 committed:
Merge pull request #25 from Wanted-Pre-Onboarding-Backend7-R/setup/#24-pr-merge-noti-and-auto-labeling

setup #24: PR Merge 알림, PR label 자동 추가

mojh7 committed:
fix #9: 다른 PR에서 MatzipRow 연결하기 위해 이름 분리

mojh7 committed:
fix #9: 다른 PR에서 MatzipRow 연결하기 위해 이름 분리

mojh7 commented:
> 장현님, 코드 잘 봤습니다!

@JoonheeJeong 

merge 전에 충돌 난 것 수정하고 이 PR에서는 MatzipRaw를 임시로 쓰고 있어서 이대로 적용하면 다른 사람들이

rebase 하고 merge할 때 고쳐야 될 수도 있으니 준희님 현재 작업중인 브랜치에서 MatzipRaw 그대로 복붙하려고 했습니다.

그런데 연관된 OpenAPIStoreXXX 관련 클래스들도 추가 해야 하더군요.

그래서 제가 이 PR에서 OpenAPIStoreXX 관련 클래스 까지 추가하는 것보단 현재 PR은 빠르게 merge될 수 있도록

데이터 파이프라인 PR merge되고 연결하는게 낫겠다 싶어서 matzip_row를 temp_matzip_row로 이름 바꿨습니다.

이름 바꿔서 jpa db 자동 생성으로 다른 분들 작업할 때 충돌날 수 있는 부분 방지하고

현재 작성된 테스트 코드는 통과되도록 했습니다.

확인해보시고 승인 해주시면 바로 merge하겠습니다 :)

---

추가로 제가 rebase 잘 못 했는지 다른 커밋들 까지 딸려왔네요 😅

https://github.com/Wanted-Pre-Onboarding-Backend7-R/tasty-finder/pull/14/commits/6751fff5651edca89c9451e332b5be8bfa79fb4e

이것만 봐주시면 됩니다

piopoi committed:
refactor #4: SecurityConfig에서 new를 이용한 객체 생성 대신 bean으로 관리되는 객체 사용하도록 수정

piopoi committed:
Merge pull request #16 from Wanted-Pre-Onboarding-Backend7-R/feat/#4-api-user-update-location

feat #4: 사용자 설정 업데이트 API 구현

SteadyFlow736 committed:
Merge pull request #17 from Wanted-Pre-Onboarding-Backend7-R/feat/#13-sgg-data-load-into-redis

feat #13: 시군구 데이터 Redis 적재 기능 구현

mojh7 committed:
fix #9: 충돌 해결
